### ISSUE TITLE
Show the checks status in the `hub pr list`

### ISSUE THREAD

**Issue by XVilka** [github/hub#2193]
Show the github checks status in the output of `hub pr list`

**Comment by joshuabezaleel** [comment]
May I have a shot on this? :smiley: @mislav

**Comment by mislav** [comment]
@joshuabezaleel For sure! The tricky part would be fetching PR statuses for all the listed PRs in a single step. (Doing a separate API request to fetch checks status for every individual PR would be too slow, IMO.) I'm not sure if batch fetching would be possible using the REST API, and instead might warrant using the GraphQL API.

**Comment by rayjlinden** [comment]
I was going to ask that all the -f options in hub ci-status be available in hub pr list.  Which I think is just an expended form of this feature request.

**Comment by mislav** [comment]
@rayjlinden It's a good idea to bring some of the fields available from `ci-status` to `pr list`. These fields would possibly have different placeholder names to fit better among the full list of fields available in `pr list --format`.

However, the most complex part of implementing this would be actually fetching the checks status for each PR listed. This should ideally be done with a single API query, since we cannot afford to make 30 separate queries to display check statuses for a list of 30 PRs.

**Comment by joshuabezaleel** [comment]
@mislav While looking at both of the API and the current CLI, I assume that the current command of `pr list` has not support filtering PR by labels?

**Comment by mislav** [comment]
> 1. I assume that the current command of `pr list` has not support filtering PR by labels?

That's true. We use [this REST endpoint](https://developer.github.com/v3/pulls/#list-pull-requests) and it doesn't offer us filtering by label. If we wanted to filter by label, we would have to switch to a REST/GraphQL Search endpoint, I think.

> 2\. the field which PullRequests can be ordered have 3 possible values which are `CREATED_AT`, `UPDATED_AT`, and `COMMENTS` (comment count) while the SORT_KEY at the current hub CLI have 4 possible values which include `long-running`. How may I resolve these different possible values?

Good find! I usually try to avoid backwards-incompatible changes, but if we gained other benefits by switching to another endpoint, I would be fine if we lost the ability to sort by `long-running`. I can't imagine it's being heavily used.

> I am new to Go and this repo so I am really sorry beforehand if I can be burdensome.

Not at all! Your questions are on-point. Thank you for taking a look!

**Comment by joshuabezaleel** [comment]
Hi @mislav , is there any example on performing request upon the GitHub GraphQL API to retrieve data in this project? Because from the discussion above I think that you suggested to use the GraphQL API and the current [`fetchPullRequests`](https://github.com/github/hub/blob/2cf4d7b5b0067f3aa89ddf9880ab4746cf923236/github/client.go#L40) method execute to the REST API v3.

**Comment by mislav** [comment]
> is there any example on performing request upon the GitHub GraphQL API to retrieve data in this project?

No, we still use REST for everything. This would be one good example where going with GraphQL might yield benefits, but we should always keep the REST functionality as fallback since older GitHub Enterprise versions (also supported by hub) don't support all the GraphQL features that GitHub.com currently supports (our GraphQL API has been evolving rapidly).

GraphQL is done by POSTing this JSON payload to `https://api.github.com/graphql`:

```json
{ "query": "...",
  "variables": { ... }
}
```

Where `query` is a GraphQL query string and `variables` is an optional dictionary of values to apply as query inputs.

The GraphQL endpoint will respond with a JSON payload as well, and that one might be tricky to parse in Go. I still haven't found an elegant way to handle GraphQL as a client in Go, and since you're just learning Go, don't feel bad if you might feel stuck at times. None if this is very easy.

**Comment by mislav** [comment]
@joshuabezaleel Definitely! Just make sure you post `query` and `variables` fields.

Here's an example GraphQL query you can try:

```graphql
fragment pr on PullRequest {
  number
  title
  commits(last: 1) {
    nodes {
      commit {
        status {
          state
        }
        checkSuites(first: 50) {
          nodes {
            conclusion
          }
        }
      }
    }
  }
}

query ($owner: String!, $repo: String!, $per_page: Int = 10) {
  repository(owner: $owner, name: $repo) {
    pullRequests(first: $per_page, orderBy: {field: CREATED_AT, direction: DESC}) {
      nodes {
        ...pr
      }
    }
  }
}
```

You will also have to use the `Accept: application/vnd.github.antiope-preview+json` request header to [gain access to the `checkSuites` functionality](https://developer.github.com/v3/checks/runs/#list-check-runs-for-a-specific-ref).

**Comment by joshuabezaleel** [comment]
Hi @mislav , I tried to run the query that you mentioned earlier at the GitHub's GraphQL Explorer and I got the error of :
```
{
  "errors": [
    {
      "path": [
        "fragment pr",
        "commits",
        "nodes",
        "commit",
        "checkSuites"
      ],
      "extensions": {
        "code": "undefinedField",
        "typeName": "Commit",
        "fieldName": "checkSuites"
      },
      "locations": [
        {
          "line": 10,
          "column": 9
        }
      ],
      "message": "Field 'checkSuites' doesn't exist on type 'Commit'"
    }
  ]
}
```
Or should this query be executed using the `PostJSON` function and not at the explorer?

**Comment by mislav** [comment]
Note what I said in my previous comment:

> You will also have to use the `Accept: application/vnd.github.antiope-preview+json` request header to [gain access to the `checkSuites` functionality](https://developer.github.com/v4/previews/#checks).

You have to include this header. The GraphQL Explorer doesn't use this header, so it can't access Checks.

**Comment by joshuabezaleel** [comment]
Hi @mislav , I am currently in the middle of this. Sorry for just getting back to you.

I want to ask what's the difference between the query that you posted here to the `mergeable` state from the field of a pullRequestConnection. I attached the screenshot below.

![image](https://user-images.githubusercontent.com/7043511/68637769-40176480-0532-11ea-9b86-e93b646ff482.png)

I tried to run the query that ask for this `mergeable` field and found that the PR with the number of [2308](https://github.com/github/hub/pull/2308) returned a CONFLICTING state (screenshot attacheded below) while at the pull request site, it shows a success state. 

![image](https://user-images.githubusercontent.com/7043511/68637954-b0be8100-0532-11ea-838b-4f19c65a3d60.png)
![image](https://user-images.githubusercontent.com/7043511/68638002-cdf34f80-0532-11ea-99f5-8efa889f70e3.png)


Thank you very much beforehand! Have a nice day! :slightly_smiling_face: 

> @joshuabezaleel Definitely! Just make sure you post `query` and `variables` fields.
> 
> Here's an example GraphQL query you can try:
> 
> ```graphql
> fragment pr on PullRequest {
>   number
>   title
>   commits(last: 1) {
>     nodes {
>       commit {
>         status {
>           state
>         }
>         checkSuites(first: 50) {
>           nodes {
>             conclusion
>           }
>         }
>       }
>     }
>   }
> }
> 
> query ($owner: String!, $repo: String!, $per_page: Int = 10) {
>   repository(owner: $owner, name: $repo) {
>     pullRequests(first: $per_page, orderBy: {field: CREATED_AT, direction: DESC}) {
>       nodes {
>         ...pr
>       }
>     }
>   }
> }
> ```
> 
> You will also have to use the `Accept: application/vnd.github.antiope-preview+json` request header to [gain access to the `checkSuites` functionality](https://developer.github.com/v4/previews/#checks).

**Comment by mislav** [comment]
> I tried to run the query that ask for this `mergeable` field and found that the PR with the number of 2308 returned a CONFLICTING state (screenshot attacheded below) while at the pull request site, it shows a success state.

"Mergeable" is completely different information than Checks status:
* Mergeable refers to the fact that the head branch of the PR can be merged cleanly to its base branch, as opposed to "conflicting", when git conflicts need to be manually resolved first before it can be merged.
* The green checkmark (or sometimes a red ❌) indicate the passing or failure state of Checks and Statuses that are active in this project. Usually this is the result of CI (continuous integration) being set up.
